### PR TITLE
feat(new transform): Aggregate transform to reduce metric volume maintaining data

### PR DESCRIPTION
**Pull request by ross**
The transform combines (`add`'s) `Incremental` metric events and uses "last one wins" semantics for `Aggregate`s. This is similar to how statsd works, but it does not (yet) support things like continuing to emit 0's for metrics it's previously seen.

* [x] Testing
* [x] TODO discussion & address
* [x] Documentation
* [x] Anything else that comes up in discussion and review

/cc https://github.com/timberio/vector/issues/676 which requested statsd functionality
/cc https://github.com/timberio/vector/issues/3668 which is functionality that could be built on top of this PR if desired.

### CODE REVIEW

**Comment by spencergilbert** [comment]
Seems reasonable to me, but I'll leave the TODO's/questions to others

File: src/transforms/aggregate.rs [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+            match event.as_metric().series().name.name.as_str() {
+                "counter_a" => assert_eq!(counter_a_summed, event),
+                "gauge_a" => assert_eq!(gauge_a_2, event),
+                _ => assert!(false),
+            };
+        }
+
+        // There were only 2
+        assert_eq!(2, count);
+    }
```
Comment (tobz) [inline src/transforms/aggregate.rs:366]: I'd personally feel better if we used a bigger interval, and then used `tokio::time::pause` and `tokio::time::advance` to more concretely demonstrate the "doesn't yield until interval elapses" behavior.

**Review by tobz — CHANGES_REQUESTED** [review]
Overall this is looking pretty good, but a few things are jumping out at the moment.

File: src/transforms/aggregate.rs [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+            for (series, metric) in map.drain() {
+                let metric = metric::Metric::from_parts(series, metric, metadata.clone());
+                output.push(Event::Metric(metric));
+                count += 1;
+            }
+        }
+
+        emit!(AggregateFlushed);
+        return count;
+    }
```
[the remark this answers is not in the document]
Comment (ross) [reply src/transforms/aggregate.rs:137]: This likely comes from my new-ness to Rust... Didn't realize that `mut` was exclusive and implied that only 1 thing could be modifying the `HashMap` at a time.  

I hadn't seen/read anything that lead me to think that `mut` provided protection nor did I find anything implying that `.drain()` atomically removes all the items in the map before setting up its iterator and returning it. The "keeps" the memory bit from https://doc.rust-lang.org/beta/std/collections/struct.HashMap.html#method.drain made me think the opposite. 

> Clears the map, returning all key-value pairs as an iterator. Keeps the allocated memory for reuse.

That lead to the a/b implementation and the TODO: question around locking. 

The `map` var used for `insert`ing is `&mut` https://github.com/timberio/vector/blob/f3118acedada401eb4749d053d8cb209c377d513/src/transforms/aggregate.rs#L82-L85

So the fact that both are `mut` implies that Rust won't allow them to happen concurrently?

File: src/transforms/aggregate.rs (hunk 1 above)
Comment (ross) [reply src/transforms/aggregate.rs:366]: Will take a look. Was working off `reduce` as an example and didn't see examples of how to muck with time in tests. Now that I have `tokio::time::pause` to grep for I found a couple examples in other places that I should be able to adapt.

Comment (ross) [reply src/transforms/aggregate.rs]: I think I addressed this in https://github.com/timberio/vector/pull/7846/commits/1da1e06ba62e0b97200946bfbb3864e1b9416a7d.

File: src/transforms/aggregate.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tobz) [reply src/transforms/aggregate.rs:137]: Yep.  One of Rust's fundamental design choices is "aliasing XOR mutability": at any given time, you can either have multiple immutable references to the same piece of data (`&foo`) or a single mutable reference (`&mut foo`), but not both unless you explicitly attempt to evade the borrow checker/compiler.

Everything else follows from that design choice WRT the "atomicity" of `HashMap::drain`, etc.

File: src/transforms/aggregate.rs [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+
+        let mut count = 0;
+        while let Some(event) = out_stream.next().await {
+            count += 1;
+            match event.as_metric().series().name.name.as_str() {
+                "counter_a" => assert_eq!(counter_a_summed, event),
+                "gauge_a" => assert_eq!(gauge_a_2, event),
+                _ => assert!(false),
+            };
+        }
```
Comment (tobz) [inline src/transforms/aggregate.rs:367]: So, this is closer, but not quite there yet.

The crucial bit here is that we actually need to poll the transform's output stream _after_ pausing time, to demonstrate that even though the input values are immediately available, our flush logic is correctly not flushing those items out because the interval hasn't yet elapsed.

We can combine the `poll!` macro from the `futures` crate and `Stream::next` to ensure that the stream does not yet have any items to return:
```rust
assert_eq!(Poll::Pending, poll!(out_stream.next()).await);
```

This code, instead of fully `await`ing for an actual value (or stream end), simply polls the stream one time to get its instantaneous state.  Given that this stream has to be polled/driven to update its internal state, we know that if we poll it once, and we get back `Poll::Pending`, that it is in the correct state.

After that, once we advance the time -- no need to `resume`, actually -- the existing `while` loop confirms that it flushes properly, and we do our remaining checks on the validity of the returned events, and so on.

**Review by tobz — CHANGES_REQUESTED** [review]
Getting closer!

File: src/transforms/aggregate.rs (hunk 3 above)
Comment (ross) [reply src/transforms/aggregate.rs:348]: Will look into this. Fwiw this test was cribbed straight from reduce's suite so if something is lacking in it the same is probably true there.

Comment (ross) [reply src/transforms/aggregate.rs:348]: The test code flushes due to the "we're shutting things down" state where `input_rx.next() => None` so it explicitly doesn't wait for the interval to elapse and the outputs are effectively there to read immediately. That seems to be the same for `reduce` and its tests which the transform function was ported from. There's no tests there of whether the `poll_period` works afaict they're all relying on the input stream has ended flush.

> The crucial bit here is that we actually need to poll the transform's output stream _after_ pausing time, to demonstrate that even though the input values are immediately available, our flush logic is correctly not flushing those items out because the interval hasn't yet elapsed.

Still looking into this. I think I'll have to manually manage and feed the input stream rather than create one from a vector to get something I can avoid seeing the end/close immediately.

Comment (ross) [reply src/transforms/aggregate.rs]: OK, pushed up a commit with tests for both the final closing down flush as well as an interval based on. The interval version is pretty involved, but after doing a bunch of tracing/debugging this seems to be what it takes to reliably get things to happen when/as expected.

File: src/transforms/aggregate.rs [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+        emit!(AggregateEventRecorded);
+    }
+
+    fn flush_into(&mut self, output: &mut Vec<Event>) -> u64 {
+        let mut count = 0_u64;
+        let map = &mut self.map;
+
+        if map.len() > 0 {
+            // TODO: not clear how this should work with aggregation so just stuffing a default one
+            // in for now.
```
Comment (tobz) [inline src/transforms/aggregate.rs:106]: Good point, I'll have to look into this.

File: src/transforms/aggregate.rs [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+
+    fn record(&mut self, event: Event) {
+        let metric = event.as_metric();
+        let series = metric.series();
+        let data = metric.data();
+        let map = &mut self.map;
+
+        match data.kind {
+            metric::MetricKind::Incremental => {
+                match map.get_mut(&series) {
```
Comment (tobz) [inline src/transforms/aggregate.rs:81]: Just directly reference the map via `self.map`. You've already got a mutable reference to it.

File: src/transforms/aggregate.rs [hunk 6]
```diff
@@ -0,0 +1,440 @@
+// let g:cargo_makeprg_params = = '--lib --no-default-features --features=transforms-aggregate transforms::aggregate'
```
Comment (tobz) [inline src/transforms/aggregate.rs:1]: Can you drop this?

**Review by tobz — CHANGES_REQUESTED** [review]
Overall, this looks good.  There's still the open question about event metadata which I'll get an answer to, but other than that, we just need to nail down the performance/polish.

File: src/transforms/aggregate.rs [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+                                    false
+                                }
+                            }
+                        }
+                    };
+                    yield stream::iter(output.into_iter());
+                    if done { break }
+                }
+            }
+            .flatten(),
```
[the remark this answers is not in the document]
Comment (ross) [reply src/transforms/aggregate.rs:161]: Yeah. I actually started down the path of rewirting this at one point b/c it was hard to follow, but then rememembered it was a direct copy-n-past from reducer so let it alone so that the implementations would match, e.g. a bug or change in one would directly apply to the other. If you don't care if they're dissimilar i can make this change. https://github.com/timberio/vector/blob/41409e7b6d1049491e1c591cbea6ba3b49222ed6/src/transforms/reduce/mod.rs#L263-L290

File: src/transforms/aggregate.rs [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+    fn flush_into(&mut self, output: &mut Vec<Event>) -> u64 {
+        let mut count = 0_u64;
+        let map = &mut self.map;
+
+        if map.len() > 0 {
+            // TODO: not clear how this should work with aggregation so just stuffing a default one
+            // in for now.
+            let metadata = EventMetadata::default();
+
+            for (series, metric) in map.drain() {
```
[the remark this answers is not in the document]
Comment (ross) [reply src/transforms/aggregate.rs:109]: K. That's just an old habit from other languages of avoiding repeated `self.`/`this->`/etc.

File: src/transforms/aggregate.rs (hunk 6 above)
Comment (ross) [reply src/transforms/aggregate.rs:1]: Doh. That was just a note to myself when I was poling around and wasn't intended to make it into a commit.

File: src/transforms/aggregate.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (tobz) [reply src/transforms/aggregate.rs]: Yeah, the implementation in reduce is just, IMO, suboptimal.  We have a chance here to improve on that original decision, so that's the driving force here.

**Review by bruceg — CHANGES_REQUESTED** [review]
Thanks for this contribution. I have a number of suggestions for improvement below. Note that all metric sinks use the metric batch buffer, which does this kind of aggregation as well. The documentation for this should indicate what this provides in comparison with that built-in buffer.

File: docs/reference/components/transforms/aggregate.cue [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+			"aarch64-unknown-linux-musl":     true
+			"armv7-unknown-linux-gnueabihf":  true
+			"armv7-unknown-linux-musleabihf": true
+			"x86_64-apple-darwin":            true
+			"x86_64-pc-windows-msv":          true
+			"x86_64-unknown-linux-gnu":       true
+			"x86_64-unknown-linux-musl":      true
+		}
+		requirements: []
+		warnings: []
```
Comment (binarylogic) [inline docs/reference/components/transforms/aggregate.cue:35]: Should there be a warning about loss of precision for any metric types?

File: docs/reference/components/transforms/aggregate.cue [hunk 10]
```diff
@@ -0,0 +1,188 @@
+package metadata
+
+components: transforms: aggregate: {
+	title: "Aggregate"
+
+	description: """
+		Aggregates multiple metric events into a single metric event based on a
+		the MetricKind. Incremental metrics are "added", Absolute uses last
+		value wins semantics.
```
Comment (binarylogic) [inline docs/reference/components/transforms/aggregate.cue:9]:
```suggestion
		Aggregates multiple metric events into a single metric event based on a defined interval window. This helps to reduce metric volume as the cost of granularity.
```

The description should be vague and the `how_it_works` section should deep dive into the details.

File: docs/reference/components/transforms/aggregate.cue [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+				},
+			]
+		},
+	]
+
+	telemetry: metrics: {
+		events_recorded_total: components.sources.internal_metrics.output.metrics.events_recorded_total
+		flushed_total:         components.sources.internal_metrics.output.metrics.flushed_total
+	}
+}
```
Comment (binarylogic) [inline docs/reference/components/transforms/aggregate.cue:188]: We need a `how_it_works` section that describes:

1. How metric types merge (absolute vs incremental, which timestamp is used, etc).
2. What the user loses when this occurs (granularity, precision, etc).
3. Why someone would do this? (cost reasons due to volume reduction).

File: src/transforms/aggregate.rs [hunk 12]
```diff
@@ -0,0 +1,420 @@
...
+    fn record(&mut self, event: Event) {
+        let metric = event.as_metric();
+        let series = metric.series();
+        let data = metric.data();
+
+        match data.kind {
+            metric::MetricKind::Incremental => {
+                match self.map.get_mut(&series) {
+                    // We already have something, add to it, will update timestamp as well.
+                    Some(existing) => existing.update(data),
```
[the remark this answers is not in the document]
Comment (ross) [reply src/transforms/aggregate.rs:81]: I chose not to explicitly handle it since there was no meaningful thing to do in that case. Only thing I can think of to do would be to emit an internal metric like `AggregateIncompatibleValue` or something of that sort.

File: src/transforms/aggregate.rs [hunk 13]
```diff
@@ -0,0 +1,420 @@
...
+                // Always replace/store
+                self.map.insert(series.clone(), data.clone());
+            }
+        };
+
+        emit!(AggregateEventRecorded);
+    }
+
+    fn flush_into(&mut self, output: &mut Vec<Event>) -> u64 {
+        let mut count = 0_u64;
```
[the remark this answers is not in the document]
Comment (ross) [reply src/transforms/aggregate.rs:99]: It was more useful in previous iterations of the code. It can definitely go now.

File: src/transforms/aggregate.rs [hunk 14]
```diff
@@ -0,0 +1,420 @@
...
+
+impl TaskTransform for Aggregate {
+    fn transform(
+        self: Box<Self>,
+        mut input_rx: Pin<Box<dyn Stream<Item = Event> + Send>>,
+    ) -> Pin<Box<dyn Stream<Item = Event> + Send>>
+    where
+        Self: 'static,
+    {
+        let mut me = self;
```
[the remark this answers is not in the document]
Comment (ross) [reply src/transforms/aggregate.rs:126]: This was copied verbatim from `reduce`.

File: src/transforms/aggregate.rs [hunk 15]
```diff
@@ -0,0 +1,420 @@
...
+
+        let interval = me.interval;
```
[the remark this answers is not in the document]
Comment (ross) [reply src/transforms/aggregate.rs:128]: This was copied verbatim from `reduce`.

File: src/transforms/aggregate.rs [hunk 16]
```diff
@@ -0,0 +1,420 @@
...
+            .with_tags(Some(tags)),
+        )
+    }
+
+    #[test]
+    fn incremental() {
+        let mut agg = Aggregate::new(&AggregateConfig { interval_ms: Some(1000_u64) }).unwrap();
+
+        let tags: BTreeMap<String, String> =
+            vec![("tag1".into(), "val1".into())].into_iter().collect();
```
[the remark this answers is not in the document]
Comment (ross) [reply src/transforms/aggregate.rs:196]: Previous generations of the code weren't using `MetricSeries` as the key and there were tests with differing tags at that point. Agree they don't add anything now.

**Review by ross — COMMENTED** [review]
> Note that all metric sinks use the metric batch buffer, which does this kind of aggregation as well. The documentation for this should indicate what this provides in comparison with that built-in buffer.

I was not aware of this and can't find any references to it in the docs. 

It seems really surprising that sinks would be modifying the data they receive and not passing it along verbatim. I think I've tracked down the code in the buffer/metrics.rf's `MetricSet` that makes that happen https://github.com/timberio/vector/blob/master/src/sinks/util/buffer/metrics.rs#L215-L218

As best as I can tell though only some of the sinks use that buffer https://github.com/timberio/vector/search?q=MetricsBuffer (note they at least mention it, didn't dig into its actual use.I spot checked a couple of their docs pages and didn't find any mention of the behavior. 

With a buffer combining it also seems like the granularity would be completely unpredictable (unless there's a clear time bound on the buffer or its flushing.) 

So I think the advantage to this transform is that it would be explicit, configurable, and consistent rather than sometimes happening behind the scenes and it would allow the functionality for any sink or any downstream transform. IMO sinks shouldn't be touching the data at all and should faithfully pass along exactly what they get. Obviously that'd be a complicated state to get to since some of them are already doing things that people may be (unknowingly) relying on.

File: Cargo.toml [hunk 17]
```diff
@@ -485,6 +485,7 @@ transforms-logs = [
 ]
 transforms-metrics = [
   "transforms-add_tags",
+  "transforms-aggregate",
```
[the remark this answers is not in the document]
Comment (bruceg) [reply Cargo.toml:488]: Absolutely. Some of the practices have changed since it was written last year, and the team has grown as well, meaning more and different reviewers. My comment on naming was more reflective of our sinks, which increasingly use distinguishing names. As you rightly point out, the `reduce` transform does not specifically indicate it operates on logs, nor do most of the transforms, so this naming is probably fine.

File: src/transforms/aggregate.rs [hunk 18]
```diff
@@ -0,0 +1,420 @@
...
+        "aggregate"
+    }
+}
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub struct Aggregate {
+    interval: Duration,
+    map: HashMap<metric::MetricSeries, metric::MetricData>,
```
[the remark this answers is not in the document]
Comment (bruceg) [reply src/transforms/aggregate.rs:61]: Yes, it is true the Datadog API keys will not be handled ideally, I don't see any preferable alternative, especially given the other metadata in use (ie acknowledgement tracking).

File: src/transforms/aggregate.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (bruceg) [reply src/transforms/aggregate.rs:126]: Interesting. AFAICT that is the only place in the code base where that exact pattern is used, although I see a few more instances of `let mut inner = self;` In any case, I would still discourage it if possible.

**Comment by bruceg** [comment]
> > Note that all metric sinks use the metric batch buffer, which does this kind of aggregation as well. The documentation for this should indicate what this provides in comparison with that built-in buffer.
> 
> I was not aware of this and can't find any references to it in the docs.

I agree, that is a definite omission that we need to address.

> It seems really surprising that sinks would be modifying the data they receive and not passing it along verbatim.

For some of the sinks, it is simply not possible to avoid. In particular, the Prometheus text and `remote_write` formats can only expose one metric per time series, and so they require some aggregation before that. The `prometheus_exporter` sink avoids using one of the standard buffers entirely because it handles it in a different way internally.

> As best as I can tell though only some of the sinks use that buffer https://github.com/timberio/vector/search?q=MetricsBuffer (note they at least mention it, didn't dig into its actual use.I spot checked a couple of their docs pages and didn't find any mention of the behavior.

There are only three metric sinks that avoid using the `MetricsBuffer`:
1. The `prometheus_exporter` sink as referenced above,
2. The `humio_metrics` sink effectively converts the incoming metrics into logs and runs a `humio_logs` sink, and
3. The `statsd` sink similarly encodes the events into its text format and buffers them as text.

> With a buffer combining it also seems like the granularity would be completely unpredictable (unless there's a clear time bound on the buffer or its flushing.)

The default configuration, controlled by `BatchConfig`, puts a bound on both the number of events (metrics) in the buffer and the amount of time in the buffer. See, for example, [the `datadog_sink` batch config](https://vector.dev/docs/reference/configuration/sinks/datadog_metrics/#batch). Raising the number of events to a large number would strictly emit events from the buffer based on the time interval.

> So I think the advantage to this transform is that it would be explicit, configurable, and consistent rather than sometimes happening behind the scenes and it would allow the functionality for any sink or any downstream transform.

I agree this transform is valuable to Vector, and I hope my discussion did not imply otherwise. The ability to use it in a transform pipeline is indeed an advantage.

File: src/transforms/aggregate.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (ross) [reply src/transforms/aggregate.rs:61]: So I was able to do this, but ran into a rust hiccup where the `.and_modify` closure is moving `metadata` and then `.or_insert` wasn't able to take ownership of it. I worked around that by doing a `.clone()` in the closure. LMK if there's a better/rustier way to go about that.

File: src/transforms/aggregate.rs (hunk 4 above)
Comment (ross) [reply src/transforms/aggregate.rs:103]: Done & TODO removed. See https://github.com/timberio/vector/pull/7846/files#r653754356 for a hiccup I ran into though.

File: src/transforms/aggregate.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (ross) [reply src/transforms/aggregate.rs:128]: Done.

File: src/transforms/aggregate.rs [hunk 19]
```diff
@@ -0,0 +1,420 @@
...
+    }
+
+    #[test]
+    fn incremental() {
+        let mut agg = Aggregate::new(&AggregateConfig { interval_ms: Some(1000_u64) }).unwrap();
+
+        let tags: BTreeMap<String, String> =
+            vec![("tag1".into(), "val1".into())].into_iter().collect();
+        let counter_a_1 = make_metric("counter_a", metric::MetricKind::Incremental,
+            metric::MetricValue::Counter { value: 42.0 }, tags.clone());
```
[the remark this answers is not in the document]
Comment (ross) [reply src/transforms/aggregate.rs:198]: It wasn't really bothering me so I haven't done this.

**Review by ross — COMMENTED** [review]
Most of the changes done. There's a couple where there's still outstanding questions before I could do anything with them. I think this is pretty shippable as-is.

**Comment by ross** [comment]
> For some of the sinks, it is simply not possible to avoid. In particular, the Prometheus text and `remote_write` formats can only expose one metric per time series, and so they require some aggregation before that. The `prometheus_exporter` sink avoids using one of the standard buffers entirely because it handles it in a different way internally.

Yeah it makes sense when there's a requirement/limitation downstream as described here, but still don't think I'd want it to be happening implicitly when not forced to. 

> I agree this transform is valuable to Vector, and I hope my discussion did not imply otherwise. The ability to use it in a transform pipeline is indeed an advantage.

:cool:

File: src/transforms/aggregate.rs [hunk 20]
```diff
@@ -0,0 +1,397 @@
...
+
+        match data.kind {
+            metric::MetricKind::Incremental => {
+                self.map.entry(series)
+                    .and_modify(|existing| {
+                        if ! existing.0.update(&data) {
+                            emit!(AggregateUpdateFailed);
+                        }
+                        existing.1.merge(metadata.clone());
+                    }).or_insert((data, metadata));
```
Comment (bruceg) [inline src/transforms/aggregate.rs:90]: The other pattern here, to avoid the `clone`, would be a match:
```suggestion
                match self.map.entry(series) {
                    hash_map::Occupied(existing) => {
                        if ! existing.0.update(&data) {
                            emit!(AggregateUpdateFailed);
                        }
                        existing.1.merge(metadata);
                    }
                    hash_map::Vacant(entry) => entry.insert((data, metadata)),
                }
```

File: src/transforms/aggregate.rs [hunk 21]
```diff
@@ -0,0 +1,397 @@
...
+                // Always replace/store
+                self.map.insert(series, (data, metadata));
+            }
+        };
+
+        emit!(AggregateEventRecorded);
+    }
+
+    fn flush_into(&mut self, output: &mut Vec<Event>) {
+        if self.map.len() > 0 {
```
Comment (bruceg) [inline src/transforms/aggregate.rs:102]: Does this condition add any value? If the `len` is zero, the map drain will be a no-op too. I would also expect clippy to gripe about this should use `is_empty`.

File: docs/reference/components/transforms/aggregate.cue [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+			title: "Aggregation Behavior"
+			body: """
+					Metrics are aggregated based their kind. During an interval `incremental` metrics
+					are "added" and `absolute` metrics use most recent value wins semantics. This results in a reduction
+					of volume, less granularity, while maintaining numerical correctness. As an example two
+					`incremental` `counter`s with values 10 and 13 processed by the transform during a period would be
+					aggregated into a single `incremental` `counter` with a value of 23. Two `absolute` `gauge`s with
+					values 93 and 95 would result in a single `absolute` `gauge` with the value of 95. More complex
+					types like `distribution`, `histogram`, `set`, and `summary` behave similarly with `incremental`
+					values being combined in a manner that makes sense based on their type.
```
Comment (bruceg) [inline docs/reference/components/transforms/aggregate.cue:195]:
```suggestion
					Metrics are aggregated based on their kind. During an interval, `incremental` metrics
					are "added" and newer `absolute` metrics replace older ones in the same series. This results in a reduction
					of volume and less granularity, while maintaining numerical correctness. As an example, two
					`incremental` `counter` metrics with values 10 and 13 processed by the transform during a period would be
					aggregated into a single `incremental` `counter` with a value of 23. Two `absolute` `gauge` metrics with
					values 93 and 95 would result in a single `absolute` `gauge` with the value of 95. More complex
					types like `distribution`, `histogram`, `set`, and `summary` behave similarly with `incremental`
					values being combined in a manner that makes sense based on their type.
```

**Review by bruceg — COMMENTED** [review]
This looks much better. Down to a couple of minor tweaks from me.

File: docs/reference/components/transforms/aggregate.cue (hunk 22 above)
Comment (ross) [reply docs/reference/components/transforms/aggregate.cue]: Done.

File: src/transforms/aggregate.rs [hunk 23]
```diff
@@ -0,0 +1,440 @@
...
+        let (series, data, metadata) = event.into_metric().into_parts();
+
+        match data.kind {
+            metric::MetricKind::Incremental => match self.map.entry(series) {
+                Entry::Occupied(mut entry) => {
+                    let existing = entry.get_mut();
+                    if !existing.0.update(&data) {
+                        emit!(AggregateUpdateFailed);
+                    }
+                    existing.1.merge(metadata);
```
Comment (bruceg) [inline src/transforms/aggregate.rs:82]: In other sinks, when the update fails, we drop the older metric instead of the newer one as is being done here. I think that would be the more sensible behavior, but I am open to your thoughts. Suggestion:
```suggestion
                    if !existing.0.update(&data) {
                        emit!(AggregateUpdateFailed);
                        *existing = (data, metadata);
                    } else {
                        existing.1.merge(metadata);
                    }
```
Note that the assignment causes an implicit drop of the metadata, which is reasonable given that we are also dropping the data as well.

**Review by bruceg — COMMENTED** [review]
I think I am good with this except for the question about handling update failures.

File: src/transforms/aggregate.rs (hunk 23 above)
Comment (ross) [reply src/transforms/aggregate.rs:82]: Ok. Didn't know the precedent and in a vacuum either option seemed equally bad so I went with the simpler one. Will update now and add in a test for the behavior.

Comment (ross) [reply src/transforms/aggregate.rs:82]: Actually @bruceg, it looks like `MetricData.update` succeeds when the kind changes (glad i tried to write tests for that case) so that code is "add"ing `Incremental` and `Absolute` values which doesn't really make sense. I looked into the metrics buffer and I think it's checking for type matches before doing the add. Kind of seems like `update` should do the add or replace logic based on `kind` so that all the users of that method don't have to check and handle that case, but for now I'll rework this to check `kind` first.

Comment (ross) [reply src/transforms/aggregate.rs]: Ok. Change made in the spirit of your request and the edge cases are now tested and behaving as expected/intended.

Comment (bruceg) [reply src/transforms/aggregate.rs]: Yes, it is a new pattern so it wouldn't have been obvious.

Good catch, yes `update` doesn't check the metric kind. It looks like the logic this transform wants and the logic for the metrics buffer is different enough that I'm not clear what the proper semantics should be. I'll have to give that more thought.

**Review by bruceg — APPROVED** [review]

File: src/transforms/aggregate.rs (hunk 23 above)
Comment (ross) [reply src/transforms/aggregate.rs]: If nothing else it could fail the update, do nothing and return false. Assuming something elsewhere isn't expecting/wanting it to add things with non-matching `kind`.

Comment (bruceg) [reply src/transforms/aggregate.rs]: Yeah, I was thinking more along the lines of `update` possibly taking the new data by value instead of reference and returning it back in the case of an error as  `Result<(), MetricData>`, to avoid needing to clone data inside update. On the other hand, the metrics buffer wants to be able to do add incrementals to both kinds, while this transform wants to replace for the case of different kinds, so that's hard to shoehorn into one method.

Comment (ross) [reply src/transforms/aggregate.rs]: Guess it's not clear to me how adding `Absolute`s makes sense given they're intended to be the value as-is, e.g. if you're going 55mph at t0 and 57mpg at t1, 112mph isn't a meaningful number. It seems like `kind` and `Counter`/`Guage` kind of overlap semantic wise and can result in internally conflicting states.

Anyway, not really related to this PR so I can let it go :grin:

Comment (bruceg) [reply src/transforms/aggregate.rs]: No, you can't add absolutes, but you can *update* an absolute to a new absolute value. You can *add* an incremental to either kind as well. Agreed it's not related to this PR.

**Review by tobz — APPROVED** [review]

**Comment by tobz** [comment]
95 comments, 36 commits... hooo boy. :)

Appreciate you sticking with us on this one. 👍🏻

**Comment by binarylogic** [comment]
Thank you @ross! This is a great contribution. As @tobz, appreciate you sticking with us.

File: docs/reference/components/transforms/aggregate.cue [hunk 24]
```diff
@@ -0,0 +1,217 @@
...
+			gauge:        true
+			histogram:    true
+			set:          true
+			summary:      true
+		}
+	}
+
+	examples: [
+		{
+			title: "Aggregate over 15 seconds"
```
Comment (jszwedko) [inline docs/reference/components/transforms/aggregate.cue:69]: Just noting that this says 15 seconds but the `interval_ms` below is 5000.
